File: packages/cli/src/monitor.ts
// `parlay monitor` implementation, extracted from index.ts.
//
// Default path is relay-backed: enroll with the central relay and exec the
// `tail -F` monitor wrapper (~1.2MB per agent) instead of an independent ~40MB
// bun poll loop. --legacy-poll keeps the old independent poller for the global
// feed or environments without the relay running.
//
// ── Deregistration watchdog (robots-ycfa) ────────────────────────────────────
// A monitor outlives whatever launched it. When the owning session dies without
// tearing down — a test fixture that spawned an agent, a harness that was
// killed — the monitor is reparented to init and keeps streaming forever, and
// nothing on the machine ever reaps it. 82 such orphans accumulated, each
// holding a bun process, a `tail -F`, and a live channel in the captain's
// panel.
//
// The server side of that fix removes the leaked channel and refuses to let its
// own poll re-create it. This is the other half: a monitor periodically asks the
// registry whether it still exists, and exits when the answer is a clear no. The
// two halves compose — prune the channel, and the process that fed it retires
// itself within a minute — so a leak self-heals instead of accumulating.
//
// Every ambiguity resolves toward STAYING ALIVE, because a monitor that quits
// while its agent is real goes registered-but-deaf (robots-dcag): a failed
// request, a non-2xx, an unparseable body, or a body that is not an array all
// reset the evidence. Only repeated, successful, well-formed answers that omit
// this agent count — the server has to say "you are gone" twice in a row.

export interface MonitorDeps {
  server: string
  exitUsage: number
  die: (msg: string, code?: number) => never
  helpWanted: (cmd: string, args: string[]) => boolean
  parseArgs: (
    cmd: string,
    args: string[],
    flags?: string[],
    valueFlags?: string[],
  ) => { positionals: string[]; opts: Record<string, string | true> }
}

/** How often the watchdog asks the registry whether this agent still exists. */
export const REGISTRY_CHECK_MS = 60_000

/**
 * Consecutive clean "you are not in the registry" answers required before a
 * monitor retires itself. Two, so a single sweep landing between an agent's
 * unregister and its re-register cannot evict a live monitor.
 */
export const MISSING_STRIKES_TO_RETIRE = 2

/**
 * Classify one registry response. Pure, so the retire/stay decision is testable
 * without a server, a clock, or a child process.
 *
 * `body` is whatever came back from GET /api/chat/agents: the parsed JSON on a
 * 2xx, or null/undefined when the request failed or could not be parsed.
 * Returns the strike count carried forward — 0 means "evidence reset".
 */
export function registryStrike(
  agent: string,
  body: unknown,
  ok: boolean,
  priorStrikes: number,
): number {
  // Anything short of a well-formed answer is not evidence of anything.
  if (!ok || !Array.isArray(body)) return 0
  const present = body.some(
    entry => typeof entry === "object" && entry !== null && (entry as { id?: unknown }).id === agent,
  )
  // An empty registry means the server just restarted and has not been
  // re-populated, not that this agent was evicted. Never a strike.
  if (body.length === 0) return 0
  return present ? 0 : priorStrikes + 1
}

/**
 * Poll the registry in the background and call `retire` once the server has
 * cleanly reported this agent missing MISSING_STRIKES_TO_RETIRE times running.
 * Returns a function that stops the watchdog.
 *
 * Set PARLAY_NO_REGISTRY_WATCHDOG=1 to disable — for a monitor deliberately
 * armed before its channel is registered, and as the escape hatch if this ever
 * misjudges a live agent.
 */
export function startRegistryWatchdog(
  server: string,
  agent: string,
  retire: () => void,
  intervalMs: number = REGISTRY_CHECK_MS,
): () => void {
  if (process.env.PARLAY_NO_REGISTRY_WATCHDOG === "1") return () => {}
  let strikes = 0
  const timer = setInterval(async () => {
    let ok = false
    let body: unknown = null
    try {
      const res = await fetch(`${server}/api/chat/agents`)
      ok = res.ok
      if (ok) body = await res.json()
    } catch { /* unreachable server — fail open, handled by registryStrike */ }
    strikes = registryStrike(agent, body, ok, strikes)
    if (strikes < MISSING_STRIKES_TO_RETIRE) return
    clearInterval(timer)
    process.stderr.write(
      `parlay monitor: '${agent}' is no longer in the server's registry — it was unregistered.\n` +
      `parlay monitor:   Retiring this monitor rather than streaming a channel nobody owns.\n` +
      `parlay monitor:   Re-arm with 'parlay listen --agent ${agent}' if this was wrong.\n`,
    )
    retire()
  }, intervalMs)
  // Never hold the process open on the watchdog alone — the child's exit is what
  // ends this command.
  timer.unref?.()
  return () => clearInterval(timer)
}

export async function runMonitor(args: string[], deps: MonitorDeps): Promise<void> {
  const { server, exitUsage, die, helpWanted, parseArgs } = deps
  if (helpWanted("monitor", args)) return
  const { opts } = parseArgs("monitor", args, ["--legacy-poll", "--notify-safe"], ["--agent"])
  const agent = opts["--agent"] as string | undefined
  // --notify-safe: cap long CHAT_MSG lines so a harness Monitor tool cannot cut
  // them mid-word silently. See tools/monitor/parlay-monitor.sh for the why.
  const notifySafe = opts["--notify-safe"] === true
  const notifyBudget = Number(process.env.PARLAY_NOTIFY_BUDGET) || 400

  // Default: relay-backed. `parlay monitor --agent <id>` enrolls with the central
  // relay and execs `tail -F` on the agent's spool — one process per agent at the
  // ~1.2MB tail floor. The relay (tools/relay/parlay-relay) must be running; a
  // single relay fans out to every agent (the N×monitor half of the 1+N split).
  if (!opts["--legacy-poll"]) {
    if (!agent) {
      return die("parlay monitor: --agent <id> is required (or use --legacy-poll for the global feed)", exitUsage)
    }
    const script = new URL("../../../tools/monitor/parlay-monitor.sh", import.meta.url).pathname
    // Bun.spawn with inherited stdio → the harness Monitor sees CHAT_MSG lines on
    // stdout exactly as before. The wrapper handles enroll + tail -F.
    const scriptArgs = ["--agent", agent, ...(notifySafe ? ["--notify-safe"] : [])]
    const proc = Bun.spawn(["bash", script, ...scriptArgs], {
      stdio: ["inherit", "inherit", "inherit"],
      env: { ...process.env, PARLAY_SERVER: server },
    })
    const stopWatchdog = startRegistryWatchdog(server, agent, () => proc.kill())
    const code = await proc.exited
    stopWatchdog()
    process.exit(code)
  }

  // Legacy independent poll loop — no relay.
  const channelParam = agent ? `&channel=${encodeURIComponent(agent)}` : ""
  let lastId = ""
  process.stderr.write(`parlay monitor (legacy poll) — server ${server}${agent ? ` channel ${agent}` : " (global)"}\n`)
  process.stderr.write(`Next (from another shell): parlay send <text...>\n`)
  while (true) {
    try {
      const res = await fetch(`${server}/api/chat/poll?after=${lastId}${channelParam}`)
      // 410 Gone: the channel was deliberately unregistered. Retrying would
      // re-create it and poll forever (robots-ycfa) — stop instead.
      if (res.status === 410) {
        process.stderr.write(
          `parlay monitor: channel '${agent}' was unregistered (410) — stopping.\n` +
          `parlay monitor:   Re-arm with 'parlay listen --agent ${agent}' if this was wrong.\n`,
        )
        return
      }
      if (!res.ok) { await Bun.sleep(2000); continue }
      const msg = await res.json() as { timeout?: boolean; id?: string; role?: string; text?: string; from?: string }
      if (msg.timeout) continue
      if (msg.id && msg.role && msg.text != null) {
        lastId = msg.id
        const fromSuffix = msg.from ? `|from:${msg.from}` : ""
        let line = `CHAT_MSG|${msg.id}|${msg.role}|${msg.text}${fromSuffix}`
        if (notifySafe && line.length > notifyBudget) {
          line = `${line.slice(0, notifyBudget)} ⟪+${line.length - notifyBudget} chars truncated for notification — run: parlay history 30 --full⟫`
        }
        process.stdout.write(`${line}\n`)
      }
    } catch {
      await Bun.sleep(3000)
    }
  }
}

File: tools/cli/internal/monitor/monitor.go
// Package monitor implements `parlay monitor` and `parlay listen`.
//
// Ported from packages/cli/src/monitor.ts and packages/cli/src/listen.ts.
//
// Default path is relay-backed: enroll with the central relay and exec the
// `tail -F` monitor wrapper (tools/monitor/parlay-monitor.sh, ~1.2MB per
// agent) instead of running an independent poll loop. This is a faithful
// shell-out port — Go spawns the exact same script the TS CLI spawns via
// Bun.spawn — not a native Go client speaking the relay's Unix-socket
// protocol directly. That tradeoff (byte-for-byte parity with the TS CLI,
// lower risk, faster to land) is a deliberate migration decision already
// made under standing authority; a native rewrite is a later, separate,
// optional ticket.
//
// --legacy-poll keeps the old independent poller for the global feed or
// environments without a relay running. monitor.ts never shelled out for
// this path either (it's a plain in-process fetch loop), so neither does
// this port — runLegacyPoll below is native Go.
package monitor

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"syscall"
	"time"

	"github.com/trillium/parlay/tools/cli/internal/args"
	"github.com/trillium/parlay/tools/cli/internal/config"
	"github.com/trillium/parlay/tools/cli/internal/help"
	"github.com/trillium/parlay/tools/cli/internal/httpc"
)

// CmdMonitor is `parlay monitor`'s entry point.
func CmdMonitor(argv []string) {
	if help.Wanted("monitor", argv) {
		return
	}
	res := args.Parse("monitor", argv, []string{"--legacy-poll", "--notify-safe", "--reap", "--apply"}, []string{"--agent"})
	agent, _ := res.String("--agent")
	notifySafe := res.Bool("--notify-safe")

	// --reap is a maintenance pass over the whole host, not a stream: it needs
	// no agent, no relay, and no enrollment. Hand it straight to the script.
	if res.Bool("--reap") {
		reapArgs := []string{"--reap"}
		if res.Bool("--apply") {
			reapArgs = append(reapArgs, "--apply")
		}
		runScript(reapArgs)
		return
	}

	if !res.Bool("--legacy-poll") {
		if agent == "" {
			httpc.Die("parlay monitor: --agent <id> is required (or use --legacy-poll for the global feed)", config.ExitUsage)
			return
		}
		runRelayMonitor(agent, notifySafe)
		return
	}

	runLegacyPoll(config.ServerURL(), agent, notifySafe)
}

// Supervision policy for the monitor script (robots-gv6t). A script run that
// survives monitorMinUptime is a healthy stream that later died; anything
// shorter is thrash, and monitorMaxRestarts consecutive thrashes end the
// supervision rather than spinning forever.
const (
	monitorMinUptime     = 2 * time.Second
	monitorMaxRestarts   = 5
	monitorRestartDelay  = time.Second
	monitorRelayReplyURL = "/api/chat/reply"
)

// runRelayMonitor runs tools/monitor/parlay-monitor.sh under bash with stdio
// inherited from this process — a harness Monitor tool sees CHAT_MSG lines on
// stdout exactly as before — and SUPERVISES it (robots-gv6t).
//
// It used to be a single `cmd.Run()` followed by `os.Exit(child's code)`, a
// faithful port of monitor.ts's `Bun.spawn` + `process.exit(code)`. That made
// the agent's only reply channel exactly as durable as one bash process:
// whatever killed it — a stray signal, a reaped child — ended the channel, and
// because `parlay listen` registers and announces BEFORE getting here, the
// panel went on showing a ready agent that could no longer be reached
// (robots-gv6t). The exit code was swallowed into a harness task-failure
// notification an agent may never read, and `exec.ExitError.ExitCode()`
// reports -1 for a signalled child, so even that notification could not name
// what happened.
//
// Now: an unexplained death is respawned, every transition is reported on
// stdout (the only stream the harness turns into an agent-visible event) as
// well as stderr, a signalled death is named, and a give-up posts the outage
// to the agent's own channel so the announce that said "listening" is
// retracted where the captain can see it.
//
// cmd.Start/Wait rather than cmd.Run so startRegistryWatchdog (watchdog.go)
// has a pid to signal. monitor.ts pairs its spawn with startRegistryWatchdog
// the same way; this path had no watchdog at all, leaving pruned channels
// with an immortal monitor (robots-jkwc).
//
// Every terminal path here leaves through httpc.Exit rather than os.Exit: it
// is the CLI's one exit hook, so it is where commandreport's
// end-of-invocation report is wired (and where testsupport.RecordingExit
// substitutes in tests). A bare os.Exit would leave `parlay monitor` — one of
// the longest-lived verbs there is — permanently "running" in the
// live-command registry until the server's reaper noticed, instead of ending
// cleanly the moment the relay script exits.
func runRelayMonitor(agent string, notifySafe bool) {
	script, err := scriptPath()
	if err != nil {
		httpc.Die(fmt.Sprintf("parlay monitor: %v", err), config.ExitRuntime)
		return
	}

	scriptArgs := []string{script, "--agent", agent}
	if notifySafe {
		scriptArgs = append(scriptArgs, "--notify-safe")
	}

	consecutiveFast := 0
	for {
		cmd := exec.Command("bash", scriptArgs...)
		cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
		cmd.Env = append(os.Environ(), "PARLAY_SERVER="+config.ServerURL())

		start := now()
		if startErr := cmd.Start(); startErr != nil {
			httpc.Die(fmt.Sprintf("parlay monitor: failed to run %s — %v", script, startErr), config.ExitRuntime)
			return
		}
		childPID := cmd.Process.Pid
		stopWatchdog := startRegistryWatchdog(agent, func() { terminateProcessTree(childPID) }, registryCheckInterval)
		runErr := cmd.Wait()
		stopWatchdog()
		uptime := now().Sub(start)

		if runErr != nil && !errors.As(runErr, new(*exec.ExitError)) {
			// Wait returned a non-exit error — unexpected after a successful
			// Start, but die rather than looping.
			httpc.Die(fmt.Sprintf("parlay monitor: failed to run %s — %v", script, runErr), config.ExitRuntime)
			return
		}

		code, signal := classifyRun(runErr)
		if !shouldRestartMonitor(code) {
			// The script refused on purpose and already said why (bad
			// usage, unreachable relay, or its own supervised give-up).
			// It will refuse identically next time, so this is terminal.
			announceStreamDown(agent, describeRun(code, signal))
			httpc.Exit(code)
			return
		}

		if uptime < monitorMinUptime {
			consecutiveFast++
		} else {
			consecutiveFast = 1
		}
		if consecutiveFast > monitorMaxRestarts {
			emitMonitorNotice("down", fmt.Sprintf(
				"monitor for '%s' died %d times in a row (%s) — GIVING UP. This agent is registered but DEAF: nothing sent to it will arrive. Re-arm with: parlay listen --agent %s",
				agent, consecutiveFast, describeRun(code, signal), agent))
			announceStreamDown(agent, describeRun(code, signal))
			httpc.Exit(config.ExitRuntime)
			return
		}

		emitMonitorNotice("respawned", fmt.Sprintf(
			"monitor for '%s' ended after %s (%s) — restarting (attempt %d/%d)",
			agent, uptime.Round(time.Second), describeRun(code, signal), consecutiveFast, monitorMaxRestarts))
		sleep(monitorRestartDelay)
	}
}

// now and sleep are indirections so the supervision loop is testable without
// real elapsed time.
var (
	now   = time.Now
	sleep = time.Sleep
)

// classifyRun turns cmd.Run()'s result into an exit status plus, when the
// child was killed rather than returning, the signal that killed it.
// exec.ExitError.ExitCode() answers -1 for a signalled child and drops the
// signal on the floor, which is precisely the information the silent-death
// report was missing.
func classifyRun(runErr error) (code int, signal syscall.Signal) {
	if runErr == nil {
		return 0, 0
	}
	var exitErr *exec.ExitError
	if !errors.As(runErr, &exitErr) {
		return config.ExitRuntime, 0
	}
	if ws, ok := exitErr.Sys().(syscall.WaitStatus); ok && ws.Signaled() {
		return exitErr.ExitCode(), ws.Signal()
	}
	return exitErr.ExitCode(), 0
}

// describeRun renders classifyRun's answer for a human.
func describeRun(code int, signal syscall.Signal) string {
	if signal != 0 {
		// Name the signal AND the shell-convention status it surfaces as, so
		// a bare "exit 144" in a harness notification is greppable back to
		// this line.
		return fmt.Sprintf("killed by %s, reported as exit %d", signal, 128+int(signal))
	}
	return fmt.Sprintf("exit %d", code)
}

// shouldRestartMonitor decides whether a finished script run is worth
// respawning. The script's two deliberate refusals — EXIT_USAGE for a bad
// invocation, EXIT_RUNTIME for an unreachable relay or its own give-up — are
// self-explained and reproducible, so retrying only spams. Everything else
// (a signal, a status nothing in the script produces) is the unexplained
// death this supervision exists for.
func shouldRestartMonitor(code int) bool {
	return code != config.ExitUsage && code != config.ExitRuntime
}

// emitMonitorNotice writes a stream-lifecycle line to BOTH stdout and stderr.
// stdout matters: a harness Monitor tool raises a notification per stdout line
// and never reads stderr, so a stderr-only warning is invisible to the very
// agent whose channel just dropped. The MONITOR| prefix is deliberately
// distinct from the relay's CHAT_MSG| lines so programmatic consumers can
// filter it — see parlay-monitor.sh's own notice().
func emitMonitorNotice(kind, text string) {
	_, _ = fmt.Fprintf(os.Stdout, "MONITOR|%s|%s\n", kind, text)
	_, _ = fmt.Fprintf(os.Stderr, "parlay monitor: %s — %s\n", kind, text)
}

// announceStreamDown retracts `parlay listen`'s "listening — monitor armed"
// announcement on the agent's own channel. Best-effort by construction: the
// stream is already dead and a server that cannot be reached must not turn
// that into a second failure.
func announceStreamDown(agent, cause string) {
	ok, reason := httpc.TryPostJSON(monitorRelayReplyURL, map[string]string{
		"agent": agent,
		"text": fmt.Sprintf(
			"monitor DOWN (%s) — this channel is no longer being read. Messages sent now will not be delivered. Re-arm with: parlay listen --agent %s",
			cause, agent),
	}, httpc.DefaultTimeout)
	if !ok {
		_, _ = fmt.Fprintf(os.Stderr, "parlay monitor: could not post the stream-down notice for '%s' — %s\n", agent, reason)
	}
}

// runScript runs parlay-monitor.sh with the given args and exits with its code.
// Used only for the --reap maintenance path (the relay-monitor path uses runRelayMonitor).
func runScript(scriptArgs []string) {
	script, err := scriptPath()
	if err != nil {
		httpc.Die(fmt.Sprintf("parlay monitor: %v", err), config.ExitRuntime)
		return
	}
	cmd := exec.Command("bash", append([]string{script}, scriptArgs...)...)
	cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
	cmd.Env = append(os.Environ(), "PARLAY_SERVER="+config.ServerURL())
	if startErr := cmd.Start(); startErr != nil {
		httpc.Die(fmt.Sprintf("parlay monitor: failed to run %s — %v", script, startErr), config.ExitRuntime)
		return
	}
	if runErr := cmd.Wait(); runErr != nil {
		var exitErr *exec.ExitError
		if errors.As(runErr, &exitErr) {
			httpc.Exit(exitErr.ExitCode())
			return
		}
		httpc.Die(fmt.Sprintf("parlay monitor: failed to run %s — %v", script, runErr), config.ExitRuntime)
		return
	}
	httpc.Exit(0)
}

// scriptPath resolves tools/monitor/parlay-monitor.sh. It prefers the name
// on PATH — same precedence as identity.ContextResetCmd, so a future ticket
// that installs it into bin/ is picked up automatically and tests can stub
// it via PATH like withFakeContextReset does — and falls back to the
// repo-relative location, the Go analogue of monitor.ts's
// `new URL("../../../tools/monitor/parlay-monitor.sh", import.meta.url)`.
// This source file lives at tools/cli/internal/monitor/monitor.go, four
// directories below the repo root, same as monitor.ts's three-up traversal
// from packages/cli/src.
func scriptPath() (string, error) {
	if abs, err := exec.LookPath("parlay-monitor.sh"); err == nil && abs != "" {
		return abs, nil
	}
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		return "", errors.New("cannot locate parlay-monitor.sh: not on PATH and own source path unavailable")
	}
	root := filepath.Join(filepath.Dir(thisFile), "..", "..", "..", "..")
	path := filepath.Join(root, "tools", "monitor", "parlay-monitor.sh")
	if _, err := os.Stat(path); err != nil {
		return "", fmt.Errorf("parlay-monitor.sh not found on PATH or at %s", path)
	}
	return path, nil
}

// runLegacyPoll is the independent poll loop with no relay — a native Go
// port of monitor.ts's `while (true) { fetch(...) }` branch. Runs until the
// process is killed, same as the TS original.
func runLegacyPoll(server, agent string, notifySafe bool) {
	notifyBudget := notifyBudgetFromEnv()
	channelParam := ""
	channelDesc := " (global)"
	if agent != "" {
		channelParam = "&channel=" + url.QueryEscape(agent)
		channelDesc = " channel " + agent
	}
	fmt.Fprintf(os.Stderr, "parlay monitor (legacy poll) — server %s%s\n", server, channelDesc)
	fmt.Fprintln(os.Stderr, "Next (from another shell): parlay send <text...>")

	lastID := ""
	for {
		res := pollOnce(server, channelParam, &lastID, notifySafe, notifyBudget, os.Stdout)
		if res.stop {
			fmt.Fprintf(os.Stderr, "parlay monitor: channel '%s' was unregistered (410) — stopping.\n", agent)
			fmt.Fprintf(os.Stderr, "parlay monitor:   Re-arm with 'parlay listen --agent %s' if this was wrong.\n", agent)
			return
		}
		if res.sleep > 0 {
			time.Sleep(res.sleep)
		}
	}
}

// pollMessage is the /api/chat/poll response shape (a subset of
// wire.ChatMessage plus the poll-specific `timeout` marker).
type pollMessage struct {
	Timeout bool    `json:"timeout"`
	ID      string  `json:"id"`
	Role    string  `json:"role"`
	Text    *string `json:"text"`
	From    string  `json:"from"`
}

// pollResult is one poll iteration's outcome: how long the caller should
// sleep before the next one — 0 when it's safe to poll again immediately (a
// message arrived, or the server reported a bare timeout) — and whether the
// loop must stop for good.
type pollResult struct {
	sleep time.Duration
	stop  bool
}

// pollOnce runs a single poll iteration. Mirrors monitor.ts's try/catch
// (network error -> sleep 3s) and !res.ok (non-2xx -> sleep 2s) branches
// exactly, plus the `msg.id && msg.role && msg.text != null` guard before
// emitting a line.
//
// 410 Gone is terminal (robots-ycfa): the channel was deliberately
// unregistered and tombstoned, and retrying would re-create it and poll
// forever — which is how 82 orphan listeners accumulated. monitor.ts stops
// on 410; this port answered it with the generic non-2xx 2s retry, so the
// Go path — the one bin/parlay actually execs — kept the leak alive
// (robots-jkwc). It is the ONLY terminal status: a 500 still retries.
func pollOnce(server, channelParam string, lastID *string, notifySafe bool, notifyBudget int, out io.Writer) pollResult {
	resp, err := httpc.Client.Get(fmt.Sprintf("%s/api/chat/poll?after=%s%s", server, *lastID, channelParam))
	if err != nil {
		return pollResult{sleep: 3 * time.Second}
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode == http.StatusGone {
		return pollResult{stop: true}
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return pollResult{sleep: 2 * time.Second}
	}

	var msg pollMessage
	if err := json.NewDecoder(resp.Body).Decode(&msg); err != nil {
		return pollResult{sleep: 3 * time.Second}
	}
	if msg.Timeout {
		return pollResult{}
	}
	if msg.ID == "" || msg.Role == "" || msg.Text == nil {
		return pollResult{}
	}

	*lastID = msg.ID
	fromSuffix := ""
	if msg.From != "" {
		fromSuffix = "|from:" + msg.From
	}
	line := fmt.Sprintf("CHAT_MSG|%s|%s|%s%s", msg.ID, msg.Role, *msg.Text, fromSuffix)
	if notifySafe && len(line) > notifyBudget {
		line = fmt.Sprintf("%s ⟪+%d chars truncated for notification — run: parlay history 30 --full⟫",
			line[:notifyBudget], len(line)-notifyBudget)
	}
	_, _ = fmt.Fprintln(out, line)
	return pollResult{}
}

// notifyBudgetFromEnv mirrors monitor.ts's
// `Number(process.env.PARLAY_NOTIFY_BUDGET) || 400`: falls back to 400 on a
// non-numeric value AND on the literal value 0 (JS `0 || 400` evaluates to
// 400, since 0 is falsy).
func notifyBudgetFromEnv() int {
	if v := os.Getenv("PARLAY_NOTIFY_BUDGET"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n != 0 {
			return n
		}
	}
	return 400
}
